src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Simplification/Simplifiers/CastSimplifier.cs

```diff
@@ -825,17 +836,75 @@ private static bool IsConditionalCastSafeToRemove(
+                ? conditionalExpression
+                : null,
+            static conditionalExpression => [conditionalExpression.WhenTrue, conditionalExpression.WhenFalse],
+            static (conditionalExpression, armExpression) =>
+            {
+                Contract.ThrowIfFalse(conditionalExpression.WhenTrue == armExpression || conditionalExpression.WhenFalse == armExpression);
+                return armExpression == conditionalExpression.WhenTrue
+                    ? conditionalExpression.WhenFalse
+                    : conditionalExpression.WhenTrue;
+            },
+            cancellationToken);
+    }
+
+    private static bool IsSwitchExpressionCaseCastSafeToRemove(
+        ExpressionSyntax castNode, SemanticModel originalSemanticModel,
+        ExpressionSyntax rewrittenExpression, SemanticModel rewrittenSemanticModel, CancellationToken cancellationToken)
+    {
+        // Defer to common helper to determine if the cast can be removed.  This unified processing of `x ? y : z` and
+        // `x switch { .. => y, .. => z, .. => w, ... }` expressions. 
+        return IsSwitchOrConditionalCastSafeToRemove(
+            castNode,
+            originalSemanticModel,
+            rewrittenExpression,
+            rewrittenSemanticModel,
+            static parentExpression => parentExpression.Parent is SwitchExpressionArmSyntax { Parent: SwitchExpressionSyntax switchExpression }
+                ? switchExpression
+                : null,
+            static switchExpression => switchExpression.Arms.SelectAsArray(a => a.Expression),
+            static (switchExpression, armExpression) =>
+            {
+                if (switchExpression.Arms.Count < 2)
+                    return null;
+
+                var arm = switchExpression.Arms.Single(a => a.Expression == armExpression);
+                var armIndex = switchExpression.Arms.IndexOf(arm);
+                return armIndex == 0
+                    ? switchExpression.Arms[1].Expression
+                    : switchExpression.Arms[armIndex - 1].Expression;
+            },
+            cancellationToken);
+    }
 
-        var parent = castExpression.WalkUpParentheses();
-        if (parent.Parent is not ConditionalExpressionSyntax originalConditionalExpression)
+    private static bool IsSwitchOrConditionalCastSafeToRemove<TConditionalOrSwitchExpression>(
+        ExpressionSyntax castNode,
+        SemanticModel originalSemanticModel,
+        ExpressionSyntax rewrittenExpression,
+        SemanticModel rewrittenSemanticModel,
+        Func<ExpressionSyntax, TConditionalOrSwitchExpression?> getConditionalOrSwitchExpression,
+        Func<TConditionalOrSwitchExpression, ImmutableArray<ExpressionSyntax>> getArmExpressions,
+        Func<TConditionalOrSwitchExpression, ExpressionSyntax, ExpressionSyntax?> getAlternativeArm,
+        CancellationToken cancellationToken)
+        where TConditionalOrSwitchExpression : ExpressionSyntax
+    {
```

Comment:
all the logic in here is what we had in IsConditionalCastSafeToRemove, but made generic so it can operate on a `a?b:c` expr vs a `a switch { ... }` expr, with helper functions to deal with conditional exprs only having two arms, which switches can have many.